Source/Core/Core/AchievementManager.cpp

```diff
@@ -111,4 +123,210 @@ AchievementManager::ResponseType AchievementManager::Request(
+{
+  if (!m_dll_enabled)
+    return;
+  RA_InitClient((HWND)main_window_handle, "Dolphin", SCM_DESC_STR);
+  RA_SetUserAgentDetail(std::format("Dolphin {} {}", SCM_DESC_STR, SCM_BRANCH_STR).c_str());
+
+  RA_InstallSharedFunctions(RACallbackIsActive, RACallbackCauseUnpause, RACallbackCausePause,
+                            RACallbackRebuildMenu, RACallbackEstimateTitle, RACallbackResetEmulator,
+                            RACallbackLoadROM);
+
+  // EE physical memory and scratchpad are currently exposed (matching direct rcheevos
+  // implementation).
+  m_memory_manager = &Core::System::GetInstance().GetMemory();
+  ReinstallMemoryBanks();
+
+  m_raintegration_initialized = true;
+
+  RA_AttemptLogin(0);
+
+  // this is pretty lame, but we may as well persist until we exit anyway
+  std::atexit(RA_Shutdown);
+}
+
+void AchievementManager::ReinstallMemoryBanks()
+{
+  if (!m_dll_enabled)
+    return;
+  RA_ClearMemoryBanks();
+  int memory_bank_size = 0;
+  if (Core::GetState() != Core::State::Uninitialized)
+  {
+    memory_bank_size = m_memory_manager->GetExRamSizeReal();
+  }
+  RA_InstallMemoryBank(0, RACallbackReadMemory, RACallbackWriteMemory, memory_bank_size);
+  RA_InstallMemoryBankBlockReader(0, RACallbackReadBlock);
+}
+
+void AchievementManager::MainWindowChanged(void* new_handle)
+{
+  if (!m_dll_enabled)
+    return;
+  if (m_raintegration_initialized)
+  {
+    RA_UpdateHWnd((HWND)new_handle);
+    return;
+  }
+
+  InitializeRAIntegration(new_handle);
+}
+
+void AchievementManager::GameChanged(bool isWii)
+{
+  if (!m_dll_enabled)
+    return;
+  ReinstallMemoryBanks();
+  if (Core::GetState() == Core::State::Uninitialized)
+  {
+    m_game_id = 0;
+    return;
+  }
+
+  //  Must call this before calling RA_IdentifyHash
+  RA_SetConsoleID(isWii ? WII : GameCube);
+
+  m_game_id = RA_IdentifyHash(m_game_hash);
+  if (m_game_id != 0)
+  {
+    RA_ActivateGame(m_game_id);
+  }
+}
+
+void AchievementManager::RAIDoFrame()
+{
+  if (!m_dll_enabled)
+    return;
+  RA_DoAchievementsFrame();
+}
+
+bool WideStringToUTF8String(std::string& dest, const std::wstring_view& str)
+{
+  int mblen = WideCharToMultiByte(CP_UTF8, 0, str.data(), static_cast<int>(str.length()), nullptr,
+                                  0, nullptr, nullptr);
+  if (mblen < 0)
+    return false;
+
+  dest.resize(mblen);
+  if (mblen > 0 && WideCharToMultiByte(CP_UTF8, 0, str.data(), static_cast<int>(str.length()),
+                                       dest.data(), mblen, nullptr, nullptr) < 0)
+  {
+    return false;
+  }
+
+  return true;
+}
+
+std::string WideStringToUTF8String(const std::wstring_view& str)
+{
+  std::string ret;
+  if (!WideStringToUTF8String(ret, str))
+    ret.clear();
+
+  return ret;
+}
+
+std::vector<std::tuple<int, std::string, bool>> AchievementManager::GetMenuItems()
+{
+  if (!m_dll_enabled)
+    return std::vector<std::tuple<int, std::string, bool>>();
+  std::array<RA_MenuItem, 64> items;
```

Comment:
I notice the API says that RA_GetPopupMenuItems must be given to a pointer with enough space for at least 32 menu items... and the API straight up doesn't have this as a constant for this for readability purposes, so it's just a magic value (cool! no footgun there). So maybe this should be turned into:

```cpp
// RA_GetPopupMenuItems requires a buffer to have space for at least 32 menu items.
static constexpr size_t MIN_NUM_MENU_ITEMS = 32;
```

and then declare this as:

```cpp
std::array<RA_MenuItem, MIN_NUM_MENU_ITEMS * 2> items;
```